eskip/eskip.go

```diff
@@ -21,6 +21,144 @@ var (
 	duplicateMethodPredicateError   = errors.New("duplicate method predicate")
 )
 
+// NewEditor creates an Editor PreProcessor, that matches routes and
+// replaces the content. For example to replace Source predicates with
+// ClientIP predicates you can use
+// --edit-route='/Source[(](.*)[)]/ClientIP($1)/', which will change
+// routes as you can see:
+//
+//        # input
+//        r0: Source("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+//        # actual route
+//        edit_r0: ClientIP("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+func NewEditor(reg *regexp.Regexp, repl []byte) *Editor {
+	return &Editor{
+		reg:  reg,
+		repl: repl,
+	}
+}
+
+type Editor struct {
+	reg  *regexp.Regexp
+	repl []byte
+}
+
+// NewClone creates a Clone PreProcessor, that matches routes and
+// replaces the content of the cloned routes. For example to migrate from Source to
+// ClientIP predicates you can use
+// --clone-route='/Source[(](.*)[)]/ClientIP($1)/', which will change
+// routes as you can see:
+//
+//        # input
+//        r0: Source("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+//        # actual route
+//        clone_r0: ClientIP("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+//        r0: Source("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+func NewClone(reg *regexp.Regexp, repl []byte) *Clone {
+	return &Clone{
+		reg:  reg,
+		repl: repl,
+	}
+}
+
+type Clone struct {
+	reg  *regexp.Regexp
+	repl []byte
+}
+
+func (e *Editor) Do(routes []*Route) []*Route {
+	if e.reg == nil {
+		return routes
+	}
+
+	for i, r := range routes {
+		rr := new(Route)
+		*rr = *r
+		if doOneRoute(e.reg, e.repl, rr) {
+			routes[i] = rr
+		}
+	}
+
+	return routes
+}
+
+func (c *Clone) Do(routes []*Route) []*Route {
+	if c.reg == nil {
+		return routes
+	}
+
+	result := make([]*Route, len(routes), 2*len(routes))
+	copy(result, routes)
+	for _, r := range routes {
+		rr := new(Route)
+		*rr = *r
+
+		rr.Id = "clone_" + rr.Id
+		predicates := make([]*Predicate, len(r.Predicates))
+		for k, p := range r.Predicates {
+			q := *p
+			predicates[k] = &q
+		}
+		rr.Predicates = predicates
+
+		filters := make([]*Filter, len(r.Filters))
+		for k, f := range r.Filters {
+			ff := *f
+			filters[k] = &ff
+		}
+		rr.Filters = filters
+
+		if doOneRoute(c.reg, c.repl, rr) {
+			result = append(result, rr)
+		}
+	}
+
+	return result
+}
+
+func doOneRoute(rx *regexp.Regexp, repl []byte, r *Route) bool {
+	if rx == nil {
+		return false
+	}
+	var changed bool
+
+	for i, p := range r.Predicates {
+		ps := p.String()
+		pss := rx.ReplaceAll([]byte(ps), repl)
+		sps := string(pss)
+		if ps == sps {
+			continue
+		}
+
+		pp, err := ParsePredicates(sps)
+		if err != nil {
+			return false
```

Comment:
this error should not really happen at this point, unless we have a bug in the eskip package (which can happen). Shall we log this error with log.Error here? (Besides the logging question, continuing is good.)

Reply:
fixed

Reply:
done